examples/vault-iam-auth/README.md

```diff
@@ -0,0 +1,109 @@
+of a custom VPC.
+
+## Running this example
+You will need to create an [Amazon Machine Image (AMI)][ami] that has Vault and Consul installed,
+which you can do using the [vault-consul-ami example][vault_consul_ami]). Each of the servers
+in this example, including the instance that is authenticating to Vault has [Dnsmasq][dnsmasq]
+installed (via the [install-dnsmasq module][dnsmasq_module]) which allows it to use the Consul
+server cluster for service discovery and thereby access Vault via DNS using the domain name
+`vault.service.consul`.
+
+### Quick start
+
+1. `git clone` this repo to your computer.
+1. Build a Vault and Consul AMI. See the [vault-consul-ami example][vault_consul_ami] documentation for
+   instructions. Make sure to note down the ID of the AMI.
+1. Install [Terraform](https://www.terraform.io/).
+1. Open `vars.tf`, set the environment variables specified at the top of the file, and fill in any other variables that
+   don't have a default. Put the AMI ID you previously took note into the `ami_id` variable.
+1. Run `terraform init`.
+1. Run `terraform apply`.
+1. Run the [vault-examples-helper.sh script][examples_helper] to
+   print out the IP addresses of the Vault server and some example commands you can run to interact with the cluster:
+   `../vault-examples-helper/vault-examples-helper.sh`.
+1. Run `curl <auth-instance-ip>:8080` to check if the client instance is fetching the secret from Vault properly
+
+
+### Vault Authentication using IAM user or role
+
+To read more about Vault IAM auth, refer to [Vault AWS Auth documentation][aws_auth].
+
+### Configuring a Vault server
+
+Before we try to authenticate, we must be sure that the Vault Server is configured
+properly and prepared to receive requests. First, we must make sure the Vault server
+has been initialized (using `vault operator init`) and unsealed (using `vault operator unseal`).
+Next, we must enable Vault to support the AWS auth method (using `vault auth enable aws`).
+Finally, we must define the correct Vault Policies and Roles to declare which EC2
+Instances will have access to which resources in Vault.
+
+Policies are rules that grant or forbid access and actions to certain paths in
+Vault. You can read more about them [here][policies_doc]. With one or more
+policies on hand, you can then finally create the authentication role.
+
+When you create a Role in Vault, you define the Policies that are attached to that
+Role, how principals who assume that Role will re-authenticate, and for how long
+tokens issued for that role will be valid.
+
+In our example we create a simple Vault Policy that allows writing and reading from
+secrets in the path `secret` namespaced with the prefix `example_`, and then create
+a Vault Role that allows authentication from all instances with a specific `ami id`.
+You can read more about role creation and check which other instance metadata you can
+use on auth [here][create_role].
```

Comment:
Be sure to mention in the docs _when_ it makes sense to use IAM Auth vs. EC2 Auth.